digdag-standards/src/main/java/io/digdag/standards/operator/td/TDOperator.java

```diff
@@ -418,7 +442,43 @@ static boolean isDeterministicClientException(Exception ex)
                     return statusCode >= 400 && statusCode < 500;
             }
         }
-        else if (ex instanceof TDClientException) {
+        return isFailedBeforeSendClientException(ex);
+    }
+
+    static boolean isNeedNotRetryException(Exception ex)
+    {
+        if (ex instanceof TDClientHttpException) {
+            int statusCode = ((TDClientHttpException) ex).getStatusCode();
+            switch (statusCode) {
+                case HttpStatus.TOO_MANY_REQUESTS_429:
+                case HttpStatus.REQUEST_TIMEOUT_408:
+                case HttpStatus.UNAUTHORIZED_401:
+                case HttpStatus.FORBIDDEN_403:
+                    return false;
+                default:
+                    // return true if 4xx
+                    return statusCode >= 400 && statusCode < 500;
+            }
+        }
+        return isFailedBeforeSendClientException(ex);
+    }
+
+    static boolean isAuthenticationErrorException(TDClientHttpException ex)
+    {
+        int statusCode = ex.getStatusCode();
+        switch (statusCode) {
+            case HttpStatus.UNAUTHORIZED_401:
+            // This is not for authentication basically, but it may be 403 for auth token error. https://tools.ietf.org/html/rfc6750
+            case HttpStatus.FORBIDDEN_403:
+                return true;
```

Comment:
I am not sure 401 and 403 are only authentication errors.
But you have investigated it and as result you set condition, it is ok.